libbpf-tools/softirqs.c

```diff
@@ -0,0 +1,258 @@
+const char argp_program_doc[] =
+"Summarize soft irq event time as histograms.\n"
+"\n"
+"USAGE: softirqs [--help] [-T] [-N] [-d] [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"    softirqss            # sum soft irq event time\n"
+"    softirqss -d         # show soft irq event time as histograms\n"
+"    softirqss 1 10       # print 1 second summaries, 10 times\n"
+"    softirqss -NT 1      # 1s summaries, nanoseconds, and timestamps\n";
+
+static const struct argp_option opts[] = {
+	{ "distributed", 'd', NULL, 0, "Show distributions as histograms" },
+	{ "timestamp", 'T', NULL, 0, "Include timestamp on output" },
+	{ "nanoseconds", 'N', NULL, 0, "Output in nanoseconds" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		env.distributed = true;
+		break;
+	case 'N':
+		env.nanoseconds = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+enum {
+	HI_SOFTIRQ = 0,
+	TIMER_SOFTIRQ = 1,
+	NET_TX_SOFTIRQ = 2,
+	NET_RX_SOFTIRQ = 3,
+	BLOCK_SOFTIRQ = 4,
+	IRQ_POLL_SOFTIRQ = 5,
+	TASKLET_SOFTIRQ = 6,
+	SCHED_SOFTIRQ = 7,
+	HRTIMER_SOFTIRQ = 8,
+	RCU_SOFTIRQ = 9,
+	NR_SOFTIRQS = 10,
+};
+
+static char *vec_names[] = {
+	[HI_SOFTIRQ] = "hi",
+	[TIMER_SOFTIRQ] = "timer",
+	[NET_TX_SOFTIRQ] = "net_tx",
+	[NET_RX_SOFTIRQ] = "net_rx",
+	[BLOCK_SOFTIRQ] = "block",
+	[IRQ_POLL_SOFTIRQ] = "irq_poll",
+	[TASKLET_SOFTIRQ] = "tasklet",
+	[SCHED_SOFTIRQ] = "sched",
+	[HRTIMER_SOFTIRQ] = "hrtimer",
+	[RCU_SOFTIRQ] = "rcu",
+};
+
+static int print_count(struct softirqs_bpf__bss *bss)
+{
+	const char *units = env.nanoseconds ? "nsecs" : "usecs";
+	__u64 count;
+	__u32 vec;
+
+	printf("%-16s %6s%5s\n", "SOFTIRQ", "TOTAL_", units);
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		count = __atomic_load_n(&bss->counts[vec], __ATOMIC_RELAXED);
+		if (count > 0)
+			printf("%-16s %11llu\n", vec_names[vec], count);
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++)
+		__atomic_store_n(&bss->counts[vec], 0, __ATOMIC_RELAXED);
```

Comment:
`__atomic_exchange_n` is what you want here to read value and atomically set it to zero at the same time

Reply:
Fixed, thanks!